File: debugger/test/src/testnativetypes.c
#include <stdlib.h>
#include <stdio.h>

typedef struct _TestStruct TestStruct;
typedef struct _FunctionStruct FunctionStruct;

typedef struct {
	int a;
	struct {
		long b;
	} foo, bar;
} Anonymous;

struct _TestStruct
{
	int a;
	long b;
	float f;
	const char *hello;
};

struct _FunctionStruct
{
	void (* foo) (int a);
};

typedef struct
{
	int a : 1;
	int b : 5;
	int c : 3;
	int d : 5;
	int e : 9;
	int f : 8;
} BitField;

typedef struct _List List;

struct _List {
	int a;
	List *next;
};

void
simple (void)
{
	int a = 5;
	long b = 7;
	float f = (float) a / (float) b;

	const char *hello = "Hello World";

	printf ("Simple: %d - %ld - %g - %s\n", a, b, f, hello);
}

void
print_test_struct (struct _TestStruct *s)
{
	printf ("Struct: %d - %ld - %g - %s\n", s->a, s->b, s->f, s->hello);
}

void
test_struct (void)
{
	struct _TestStruct s;

	s.a = 5;
	s.b = 7;
	s.f = (float) s.b / (float) s.a;
	s.hello = "Hello World";

	print_test_struct (&s);
}

void
test_struct_2 (void)
{
	TestStruct s;

	s.a = 5;
	s.b = 7;
	s.f = (float) s.b / (float) s.a;
	s.hello = "Hello World";

	print_test_struct (&s);
}

void
test_struct_3 (void)
{
	Anonymous s;

	s.a = 5;
	s.foo.b = 800;
	s.bar.b = 9000;
	printf ("Test: %d - %ld,%ld\n", s.a, s.foo.b, s.bar.b);
}

void
test_func (int a)
{
	printf ("Test: %d\n", a);
}

void
test_function_struct (void)
{
	FunctionStruct test;

	test.foo = test_func;
}

void
test_bitfield (void)
{
	BitField bitfield;

	bitfield.a = 1;
	bitfield.b = 3;
	bitfield.c = 4;
	bitfield.d = 9;
	bitfield.e = 15;
	bitfield.f = 8;

	printf ("Bitfield: %x\n", * ((int *) &bitfield));
}

void
test_list (void)
{
	List list;

	list.a = 9;
	list.next = &list;

	printf ("%d\n", list.next->a);
}

int
main (void)
{
	setbuf (stdout, NULL);
	simple ();
	test_struct ();
	test_struct_2 ();
	test_struct_3 ();
	test_function_struct ();
	test_bitfield ();
	test_list ();
	return 0;
}

File: test/testsuite/testnativetypes.cs
using System;
using NUnit.Framework;

using Mono.Debugger;
using Mono.Debugger.Languages;
using Mono.Debugger.Frontend;

namespace Mono.Debugger.Tests
{
	[TestFixture]
	public class testnativetypes : TestSuite
	{
		public testnativetypes ()
			: base ("testnativetypes", "testnativetypes.c")
		{ }

		[Test]
		[Category("Native")]
		[Category("NativeTypes")]
		public void Main ()
		{
			Process process = Start ();
			Assert.IsTrue (process.MainThread.IsStopped);

			Thread thread = process.MainThread;

			AssertStopped (thread, "main", "main");

			AssertExecute ("continue");
			AssertHitBreakpoint (thread, "simple", "simple");

			AssertPrint (thread, "a", "(int) 5");
			AssertPrint (thread, "b", "(long int) 7");
			AssertPrint (thread, "f", "(float) 0.7142857");
			AssertPrint (thread, "hello", "(char *) \"Hello World\"");

			AssertExecute ("continue");
			AssertTargetOutput ("Simple: 5 - 7 - 0.714286 - Hello World");
			AssertHitBreakpoint (thread, "struct", "test_struct");

			AssertPrint (thread, "s",
				     "(_TestStruct) { a = 5, b = 7, f = 1.4, " +
				     "hello = \"Hello World\" }");
			AssertPrint (thread, "s.hello", "(char *) \"Hello World\"");

			AssertExecute ("continue");
			AssertTargetOutput ("Struct: 5 - 7 - 1.4 - Hello World");
			AssertHitBreakpoint (thread, "struct2", "test_struct_2");

			AssertPrint (thread, "s",
				     "(TestStruct) { a = 5, b = 7, f = 1.4, " +
				     "hello = \"Hello World\" }");
			AssertPrint (thread, "s.hello", "(char *) \"Hello World\"");

			AssertExecute ("continue");
			AssertTargetOutput ("Struct: 5 - 7 - 1.4 - Hello World");
			AssertHitBreakpoint (thread, "struct3", "test_struct_3");

			AssertPrint (thread, "s.a", "(int) 5");
			AssertPrint (thread, "s.foo", "() { b = 800 }");
			AssertPrint (thread, "s.bar", "() { b = 9000 }");

			AssertExecute ("continue");
			AssertTargetOutput ("Test: 5 - 800,9000");
			AssertHitBreakpoint (thread, "function struct", "test_function_struct");

			AssertExecute ("continue");
			AssertHitBreakpoint (thread, "bitfield", "test_bitfield");

			AssertPrint (thread, "bitfield",
				     "(BitField) { a = 1, b = 3, c = 4, d = 9, e = 15, f = 8 }");

			AssertExecute ("continue");
			AssertTargetOutput ("Bitfield: 3");
			AssertHitBreakpoint (thread, "list", "test_list");

			AssertPrint (thread, "list.a", "(int) 9");
			AssertPrint (thread, "list.next->a", "(int) 9");

			AssertExecute ("continue");
			AssertTargetOutput ("9");

			AssertHitBreakpoint (thread, "funcptr", "test_function_ptr");

			AssertType (thread, "func_ptr", "void (*) (int)");
			AssertType (thread, "func_ptr2", "typedef test_func_ptr = void (*) (int)");
			AssertPrintException (thread, "*func_ptr",
					     "Expression `func_ptr' is not a pointer.");
			AssertPrintException (thread, "*func_ptr2",
					      "Expression `func_ptr2' is not a pointer.");
			AssertTypeException (thread, "*func_ptr",
					     "Expression `func_ptr' is not a pointer.");
			AssertTypeException (thread, "*func_ptr2",
					     "Expression `func_ptr2' is not a pointer.");
			AssertType (thread, "func_ptr3", "test_func_ptr*");
			AssertType (thread, "*func_ptr3", "typedef test_func_ptr = void (*) (int)");
			AssertTypeException (thread, "**func_ptr3",
					     "Expression `*func_ptr3' is not a pointer.");

			AssertExecute ("continue");
			AssertTargetOutput ("Test: 3");
			AssertTargetOutput ("Test: 9");
			AssertTargetOutput ("Test: 11");

			AssertHitBreakpoint (thread, "array", "test_array");
			AssertPrint (thread, "array->simple_array", "(int []) [ 8192, 55, 71 ]");
			AssertPrint (thread, "array->multi_array",
				     "(long int []) [ [ 16, 32, 64 ], [ 24, 48, 96 ] ]");
			AssertPrint (thread, "array->anonymous_array", "(float []) [ ]");
			AssertPrint (thread, "array->anonymous_array [0]", "(float) 3.141593");
			AssertPrint (thread, "array->anonymous_array [1]", "(float) 2.718282");
			AssertExecute ("continue");

			AssertTargetExited (thread.Process);
		}
	}
}
